File: random-pick-index/random-pick-index.java
class Solution {

    public Random r;
    public HashMap <Integer, ArrayList <Integer>> map;
    
    public Solution (int[] nums) {
        map = new HashMap ();
        r = new Random ();
        for (int i = 0; i < nums.length; i++) {
            ArrayList <Integer> list = new ArrayList ();
            if (map.containsKey (nums [i]))
                list = map.get (nums [i]);
            list.add (i);
            map.put (nums [i], list);
        }
    }
    
    public int pick (int target) {
        ArrayList <Integer> list = map.get (target);
        return list.get (r.nextInt (list.size()));
    }
}
/**
 * Your Solution object will be instantiated and called as such:
 * Solution obj = new Solution(nums);
 * int param_1 = obj.pick(target);
 */

File: 398-random-pick-index/random-pick-index.js
/**
 * @param {number[]} nums
 */
var Solution = function(nums) {
   this.nums = nums;
   this.length = nums.length;
};

Solution.prototype.pick = function(target) {
    var numsArray = this.nums;
    var numsLength = this.length;
    
    if(numsLength == 1) return 0;
       
    var output = 0;
    var count = 0;
  
    
    for(var i=0; i < numsLength; i++){
        if(numsArray[i] == target && Math.floor(Math.random() * (++count)) == 0){
              output = i;
         }
    }
    
    return output;
};

/** 
 * Your Solution object will be instantiated and called as such:
 * var obj = new Solution(nums)
 * var param_1 = obj.pick(target)
 */